### PR TITLE
Fault injection tests for pooling scenarios

### PR DESCRIPTION
**Pull request by abhipsaMisra**
## Checklist
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [x] This pull-request is submitted against the `master` branch.

## Description of the changes

The following pooling fault injection tests are implemented:

Operations | Amqp/ Amqp_Ws | Comments
-- | -- | --
Send Event | KillTcp | Ok
  | ShutdownAmqp | Ok
  | KillAmqpConnection | Ok
  | KillAmqpSession | #950
  | KillAmqpD2CLink | #950
  | InvokeAuthError |  
  | InvokeMaxMessageQuota |  
  | InvokeThrottling | Ok
  | InvokeThrottling -   Long |  
  |   | 
Receive Event | KillTcp | Ok
  | ShutdownAmqp | Ok
  | KillAmqpConnection | Ok
  | KillAmqpSession | #942, #950 
  | KillAmqpC2DLink | #942
  |   |  
Set Receive   Method | KillTcp | Ok
  | ShutdownAmqp | Ok
  | KillAmqpConnection | Ok
  | KillAmqpSession | #950
  | KillAmqpMethodReqLink | #950
  | KillAmqpMethodRespLink | #950
  |   |  
Update   Reported Properties | KillTcp | Ok
  | ShutdownAmqp | Ok
  | KillAmqpConnection | Ok
  | KillAmqpSession | #950
  | KillAmqpTwinLinkReq | #950
  | KillAmqpTwinLinkResp | #950
  |   |  
Set Desired   Property Callback | KillTcp | Ok  
  | ShutdownAmqp | Ok 
  | KillAmqpConnection | Ok 
  | KillAmqpSession | #950  
  | KillAmqpTwinLinkReq | #950  
  | KillAmqpTwinLinkResp | #950  

## Following open TODO items:
#942 - NullReferenceException is thrown while device client completes Receive operation
#943  - Honor different pool sizes for different connection pool settings (this is allow for the following e2e test parallel runs: single connection/ multiple connection pools; pooled fault injection tests).
#950 - Link/session faults for message send/ method/ twin operations closes the connection

### CODE REVIEW

**Comment by CIPop** [comment]
@abhipsaMisra looks like the tests are failing. Is there an update?

File: build.ps1 [hunk 1]
```diff
@@ -284,10 +305,16 @@ try {
         if (IsWindowsDevelopmentBox)
         {
             RunTests e2e\test "End-to-end tests (NetCoreApp2.1, NET47, NET451)"
+
+            # To exclude the Pooling Fault Injection Tests from E2E test run (Pooling Fault Injection E2E tests require an additional IoTHub):
+            # RunTests e2e\test "End-to-end tests (NetCoreApp2.1, NET47, NET451)" "*" "TestCategory!=IoTHub-FaultInjection-PoolAmqp"
```
Comment (CIPop) [inline build.ps1]: Why are these commented out?

File: common/test/TestDevice.cs [hunk 2]
```diff
@@ -32,19 +32,21 @@ public class TestDevice
...
+            _iotHubConnectionString = iotHubConnectionString;
         }
 
         /// <summary>
         /// Factory method.
         /// </summary>
         /// <param name="namePrefix"></param>
         /// <param name="type"></param>
-        public static async Task<TestDevice> GetTestDeviceAsync(string namePrefix, TestDeviceType type = TestDeviceType.Sasl)
+        public static async Task<TestDevice> GetTestDeviceAsync(string namePrefix, TestDeviceType type = TestDeviceType.Sasl, bool usePrimaryHub = true)
```
Comment (CIPop) [inline common/test/TestDevice.cs]: What does it mean to use "primary hub"? We should only test against a single hub.

File: e2e/test/FaultInjectionPoolAmqpTests.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
```
Comment (CIPop) [inline e2e/test/FaultInjectionPoolAmqpTests.cs]: Missing header.

Comment (CIPop) [inline e2e/test/FaultInjectionPoolAmqpTests.cs]: Please name the file <className>.<specifier>.cs (in this case specifier could be `base`?)

File: e2e/test/MessageReceiveFaultInjectionPoolAmqpTests.cs [hunk 4]
```diff
@@ -0,0 +1,640 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (CIPop) [inline e2e/test/MessageReceiveFaultInjectionPoolAmqpTests.cs]: See above: name of the file needs to be the name of the class. We can add a .<specifier>.cs after.

File: e2e/test/MessageSendFaultInjectionPoolAmqpTests.cs [hunk 5]
```diff
@@ -0,0 +1,742 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (CIPop) [inline e2e/test/MessageSendFaultInjectionPoolAmqpTests.cs]: Same as everywhere else: please rename the files.

File: e2e/test/PoolingOverAmqp.cs [hunk 6]
```diff
@@ -9,26 +9,26 @@
...
-        public const int MuxWithoutPoolingPoolSize = 1; // For enabling multiplexing without pooling, the pool size needs to be set to 1
-        public const int MuxWithPoolingDevicesCount = 4;
-        public const int MuxWithPoolingPoolSize = 2;
+        public const int SingleConnection_DevicesCount = 2;
+        public const int SingleConnection_PoolSize = 1;
+        public const int MultipleConnections_DevicesCount = 4;
+        public const int MultipleConnections_PoolSize = 2;
         public const int maxTestRunCount = 5;
-        public const float testSuccessRate = 0.8f; // 4 out of 5 test runs should pass (even after accounting for network instability issues).
+        public const double testSuccessRate = 0.8d; // 4 out of 5 test runs should pass (even after accounting for network instability issues).
```
Comment (CIPop) [inline e2e/test/PoolingOverAmqp.cs]: Don't use double. Use int instead.

File: e2e/test/PoolingOverAmqp.cs [hunk 7]
```diff
@@ -102,7 +102,7 @@ public static class MultiplexingOverAmqp
                         Assert.AreEqual(ConnectionStatusChangeReason.Client_Close, amqpConnectionStatuses[i].LastConnectionStatusChangeReason, $"The actual connection status change reason is = {amqpConnectionStatuses[i].LastConnectionStatusChangeReason}");
                     }
                     if (deviceConnectionStatusAsExpected) successfulRuns++;
-                    reRunTest = (successfulRuns / totalRuns) < testSuccessRate;
+                    reRunTest = (((double)successfulRuns / totalRuns) * 100) < (testSuccessRate * 100);
```
Comment (CIPop) [inline e2e/test/PoolingOverAmqp.cs]: Should work fine with int if the multiplication with 100 is done first.

File: e2e/test/prerequisites/iot_config.cmd_template [hunk 8]
```diff
@@ -38,6 +27,20 @@ set IOTHUB_EVENTHUB_COMPATIBLE_NAME=
 @REM EventHub consumer group:
 set IOTHUB_EVENTHUB_CONSUMER_GROUP=
 
+@REM ---------- Secondary IoTHub configuration for running Polling Fault Injection tests ----------
```
Comment (CIPop) [inline e2e/test/prerequisites/iot_config.cmd_template]: Why do we need this? Why not just run against the primary hub?
This makes it very difficult for 3rd parties to run our tests.

File: build.ps1 (hunk 1 above)
Comment (abhipsaMisra) [reply build.ps1]: We are back to running all non-pooled-fault injection tests and then pooled-fault-injection tests.

File: common/test/TestDevice.cs (hunk 2 above)
Comment (abhipsaMisra) [reply common/test/TestDevice.cs]: This was to test if using multiple hubs brings down the run time (with current implementation). It was not found to be beneficial. This has been reverted.

File: e2e/test/prerequisites/iot_config.cmd_template (hunk 8 above)
Comment (abhipsaMisra) [reply e2e/test/prerequisites/iot_config.cmd_template]: Same as above. (https://github.com/Azure/azure-iot-sdk-csharp/pull/941#discussion_r291412800)

**Comment by abhipsaMisra** [comment]
The PR is reusing the same operations that are used for single device/ basic pooled scenarios. The test templates used for basic pooled tests and fault injection pooled tests are different.

**Pull request merged**